### DISCUSSION TITLE
How to declare authentication for OpenAPI spec and how to set per-route sources

### DISCUSSION THREAD

**Discussion by teaishealthy** [RobinTail/express-zod-api#521] [Q&A] [answered]
Hey!
1. Can I add a global or per-route authentication requirements to be added to the OpenAPI spec
2. Can I set per-route i/o sources

If 2 is currently not possible, I had imagined something similar to this:
```js
defaultEndpointsFactory.build({
    ...,
    input: {
        body: z.object({
            hello: z.string()
        })
    }
})
```

**Comment by RobinTail** [comment]
Hello @teaishealthy ,

Speaking of Authentication requirements, I presume you're talking about these:
https://swagger.io/docs/specification/authentication/

Am I correct?

This feature is currently not implemented.
But if/when it will be, I think it will be a property of the `Middleware` entity.

Regarding the input sources.
Currently you can only configure them globally. 
Perhaps I could make it also configurable on the `Endpoint` level, but not the way you described.
It could be a property sibling to `input`, not its child definitely.
Combining the input sources into the single `input` object is the concept feature of the library.

Anyway, thank you for interesting ideas. I'm going to think them through.

**Reply by teaishealthy** [reply]
Yeah, the that's those are auth requirements I was talking about.

Your ideas for the configurable input source seem reasonable (and in hindsight better than mine).

Thank you for your quick response!

**Comment by RobinTail** [comment]
Hello @teaishealthy ,

I prepared the solution for the feature you need.
Please check it out here:
https://github.com/RobinTail/express-zod-api/pull/523
and let me know if it would fulfill your needs.

The implementation does not yet support scopes yet (for OpenID and oAuth2 Authentication schemas).

**Reply by teaishealthy** [reply]
Hey there,

I've added the security option to the middleware which handles auth validation, which does work, but it shows one authorization option for each route that uses that middleware.


This is what it looks like in Swagger
![swagger](https://user-images.githubusercontent.com/76410798/183266953-d72ab11e-4eb0-4fa3-b2d4-044b05cdfbcf.png)

Here in Elements
![elements](https://user-images.githubusercontent.com/76410798/183266822-382d579a-7254-4794-96ca-8ae7ba7c2a2b.png)

The current implementation does not seem quite right as it should be one authorization field for all routes, as they use the same middleware.

Other than that it seems perfect for my use case.

**Reply by RobinTail** [reply]
Oh yes, sorry, forgot to deduplicate them ¯\_(ツ)_/¯ 
I'll fix it tomorrow.

**Reply by RobinTail** [reply]
@teaishealthy , I fixed the issue with the naming. Could you please check it again?

**Reply by teaishealthy** [reply]
Just tried it again and everything works without any problems. Thank you for the implementation and quick fix!

**Comment by RobinTail** [comment, marked as answer]
🚀  The feature has just been released in v7.7.0, @teaishealthy
